clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4212,6 +4259,50 @@ class FreeFunctionKernelBodyCreator : public SyclKernelFieldHandler {
     return DRE;
   }
 
+  Expr *createGetAddressOf(Expr *E) {
+    return UnaryOperator::Create(
+        SemaSYCLRef.getASTContext(), E, UO_AddrOf,
+        SemaSYCLRef.getASTContext().getPointerType(E->getType()), VK_PRValue,
+        OK_Ordinary, SourceLocation(), false,
+        SemaSYCLRef.SemaRef.CurFPFeatureOverrides());
+  }
+
+  Expr *createDerefOp(Expr *E) {
+    return UnaryOperator::Create(SemaSYCLRef.getASTContext(), E, UO_Deref,
+                                 E->getType()->getPointeeType(), VK_LValue,
+                                 OK_Ordinary, SourceLocation(), false,
+                                 SemaSYCLRef.SemaRef.CurFPFeatureOverrides());
+  }
+
+  Expr *createReinterpretCastExpr(Expr *E, QualType To) {
+    return CXXReinterpretCastExpr::Create(
+        SemaSYCLRef.getASTContext(), To, VK_PRValue, CK_BitCast, E,
+        /*Path=*/nullptr,
+        SemaSYCLRef.getASTContext().getTrivialTypeSourceInfo(To),
+        SourceLocation(), SourceLocation(), SourceRange());
+  }
+
+  Expr *createStructTemporary(ParmVarDecl *OrigFunctionParameter) {
+    Expr *DRE = createParamReferenceExpr();
+
+    assert(OrigFunctionParameter && "no parameter?");
+
+    CXXRecordDecl *RD = OrigFunctionParameter->getType()->getAsCXXRecordDecl();
+    InitializedEntity Entity = InitializedEntity::InitializeParameter(
+        SemaSYCLRef.getASTContext(), OrigFunctionParameter);
+
+    if (RD->hasAttr<SYCLGenerateNewTypeAttr>()) {
+      DRE = createReinterpretCastExpr(
+          createGetAddressOf(DRE), SemaSYCLRef.getASTContext().getPointerType(
+                                       OrigFunctionParameter->getType()));
+      DRE = createDerefOp(DRE);
+    }
+
+    ExprResult ArgE = SemaSYCLRef.SemaRef.PerformCopyInitialization(
```

Comment:
Can we call copy constructor? What happens if there is a user provided one ?

Reply:
Actually what happens in the non-free-function case now? I know we do a list initialization of the kernel object. Does this end up calling the copy constructor?

Reply:
We call copy constructor too. For the code:
```
#include "sycl.hpp"
sycl::queue q;

struct S {
  int a;
 };

int main() {
  S s;
  q.submit([&](sycl::handler &h) {
    h.single_task<class fake_kernel>([=]() { (void)s; });
  });

}
```
We generate
```
`-FunctionDecl 0xa210cf0 <col:38, <invalid sloc>> col:38 _ZTSZZ4mainENKUlRN4sycl3_V17handlerEE_clES2_E11fake_kerne
l 'void (S)'                                                                                                      
  |-ParmVarDecl 0xa210f20 <<invalid sloc>> <invalid sloc> used __arg_s 'S'                                        
  |-CompoundStmt 0xa2110a8 <<invalid sloc>>                                                                       
  | |-DeclStmt 0xa210e80 <./Inputs/sycl.hpp:311:23>                                                               
  | | `-VarDecl 0xa210e18 <test-no-free-func.cpp:11:38, ./Inputs/sycl.hpp:311:23> test-no-free-func.cpp:11:38 used
 __SYCLKernel '(lambda at test-no-free-func.cpp:11:38)' cinit                                                     
  | |   `-InitListExpr 0xa210e98 <./Inputs/sycl.hpp:311:23> '(lambda at test-no-free-func.cpp:11:38)'             
  | |     `-CXXConstructExpr 0xa210fc0 <col:23> 'S' 'void (const S &) noexcept'    <- HERE          
  | |       `-ImplicitCastExpr 0xa210fa8 <col:23> 'const S' lvalue <NoOp>                                         
  | |         `-DeclRefExpr 0xa210f88 <col:23> 'S' lvalue ParmVar 0xa210f20 '__arg_s' 'S'                         

```
Because we use `InitializationKind::CreateCopy` in the body creator.

We should be safe when doing so, since SYCL spec requires kernel arguments to be [device copyable](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec::device.copyable), and device copyable requirements include

- Type T has at least one eligible copy constructor, move constructor, copy assignment operator, or move assignment operator;

- Each eligible copy constructor, move constructor, copy assignment operator, and move assignment operator is public;

- When doing an inter-device transfer of an object of type T, the effect of each eligible copy constructor, move constructor, copy assignment operator, and move assignment operator is the same as a bitwise copy of the object;